File: libs/power-log-parser/src/lib/handlers/full-entity-handler.ts
import type { Helper } from '../helper';
import { Action, FullEntity, Game, Node, NodeType, Tag } from '../models';
import { Regexes } from '../regexes';
import type { ParserState } from '../state/parser-state';

export class FullEntityHandler {
	static Handle(timestamp: string, data: string, state: ParserState, indentLevel: number, helper: Helper): boolean {
		let match = Regexes.ActionFullEntityUpdatingRegex.exec(data);
		if (!match) {
			match = Regexes.ActionFullEntityCreatingRegex.exec(data);
		}
		if (match) {
			const rawEntity = match[1];
			const cardId = match[2] ?? '';
			const entity = helper.ParseEntity(rawEntity);
			state.GameState.UpdateEntityName(rawEntity);

			const fullEntity = new FullEntity();
			fullEntity.CardId = cardId;
			fullEntity.Id = entity;
			fullEntity.Tags = [];
			fullEntity.TimeStamp = timestamp;
			fullEntity.SubSpellInEffect = state.CurrentSubSpell?.GetActiveSubSpell() ?? null;

			state.UpdateCurrentNode(NodeType.Game, NodeType.Action);

			const newNode = new Node(NodeType.FullEntity, fullEntity, indentLevel, state.Node, data);
			if (state.Node!.Type === NodeType.Game) {
				(state.Node!.Object as Game).AddData(fullEntity);
			} else if (state.Node!.Type === NodeType.Action) {
				(state.Node!.Object as Action).Data.push(fullEntity);
			} else {
				throw new Error('Invalid node ' + state.Node!.Type);
			}
			state.RegisterEntityForIndex(fullEntity);
			state.CreateNewNode(newNode);
			state.Node = newNode;
			return true;
		}
		return false;
	}
}

File: tools/linux-helpers/hs-converter/HearthstoneReplays/Parser/Handlers/DataHandler/FullEntityHandler.cs
using HearthstoneReplays.Enums;
using HearthstoneReplays.Parser.ReplayData.Entities;
using HearthstoneReplays.Parser.ReplayData.GameActions;
using HearthstoneReplays.Parser.ReplayData;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Action = HearthstoneReplays.Parser.ReplayData.GameActions.Action;

namespace HearthstoneReplays.Parser.Handlers
{
    internal class FullEntityHandler
    {
        public static bool HandleFullEntity(DateTime timestamp, string data, ParserState state, int indentLevel, Helper helper)
        {
            var match = Regexes.ActionFullEntityUpdatingRegex.Match(data);
            if (!match.Success)
            {
                match = Regexes.ActionFullEntityCreatingRegex.Match(data);
            }
            if (match.Success)
            {
                var rawEntity = match.Groups[1].Value;
                var cardId = match.Groups[2].Value;
                var entity = helper.ParseEntity(rawEntity);
                state.GameState.UpdateEntityName(rawEntity);

                var fullEntity = new FullEntity { CardId = cardId, Id = entity, Tags = new List<Tag>(), TimeStamp = timestamp };
                fullEntity.SubSpellInEffect = state.CurrentSubSpell?.GetActiveSubSpell();
                //state.GameState.FullEntity(fullEntity, false);

                state.UpdateCurrentNode(typeof(Game), typeof(Action));

                var newNode = new Node(typeof(FullEntity), fullEntity, indentLevel, state.Node, data);
                if (state.Node.Type == typeof(Game))
                    ((Game)state.Node.Object).AddData(fullEntity);
                else if (state.Node.Type == typeof(Action))
                    ((Action)state.Node.Object).Data.Add(fullEntity);
                else
                    throw new Exception("Invalid node " + state.Node.Type);
                state.CreateNewNode(newNode);
                state.Node = newNode;
                return true;
            }
            return false;
        }
    }
}
